### PR TITLE
Add the option for a sentinel file for the origin

### LINKED ISSUE #995: Add support for the XRootD sentinel file
@biozit reports one useful feature from the legacy OSDF is the ability to have a "sentinel file".

The sentinel file is a file that must exist in order for the service/container to startup; if missing, then the service should shutdown with a recognizable error.  This is useful in cases where the admin may forget to mount the filesystem inside the origin (leading to a confusing set of responses due to an empty directory) or if there's a broken filesystem used by the cache

Fabio writes:

> This option checks the existence of a specific file on a XrootD volume; if the file is not present, the XrootD won't start.
> It was tested using this option:
>    `oss.space public $(rootdir) chkmount thisfileshoulbeheretoxrootdststart -`
> https://xrootd.slac.stanford.edu/doc/dev54/ofs_config.htm#_Toc89982406.

I believe we could implement either at the XRootD or the Pelican layer.  Doing this in Pelican may provide the ability to do a higher-quality error message.

In terms of configuration -- I would suggest a sentinel could be used for each of the origin's exports.

### PR DESCRIPTION
**Pull request by haoming29**
Closes #995

### CODE REVIEW

**Comment by haoming29** [comment]
This PR added an additional field `SentinelFile` to the `Origin.Exports` parameter. @CannonLock you might need to update the UI as well.

**Comment by CannonLock** [comment]
@haoming29 I will patch on the ui change after the name is decided.

**Comment by haoming29** [comment]
Trying to understand the intension for S3 here. Will the user provide a path to the sentinel file for either S3 and POSIX? Or only a file name? The latter is what I assumed in this PR and I actually checked and require that the parameter does not contain a directory.

**Comment by jhiemstrawisc** [comment]
I think you'll have to think about a good way to incorporate the origin's configured storage type. In the case of an S3 sentinel object, we might require that for a given `s3endpoint`, `bucket` and `object`, we can fetch `<s3endpoint>/<bucket>/<object>` if the origin is using path-style requests or `<bucket>.<s3endpoint>/<object>` if we're using virtual hosting.

Of course, if you want to check that an authenticated S3 object exists _before_ starting xrootd, you'll also need to handle creating the signed AWSv4 URLs with the credentials Pelican knows about, and that's not a small task. You could either write all of the functionality to do this in Pelican (and wind up duplicating half the magic of the S3 plugin) or you could use the AWS SDK and add a lot of bloat...

Maybe the approach is to start XRootD first and immediately try to pull the file/object. If the origin can't get something from itself, we assume the sentinel doesn't exist.

**Comment by haoming29** [comment]
@jhiemstrawisc That clarifies things a lot! I like your suggestion as to start xrootd first then pull a file and see if we can make it for S3 backend. For POSIX backend, I think it's easy enough to `stat` the file before starting the xrootd and that way we eliminate other possibilities that can cause the error getting the file through the origin (auth/token/etc). I will incorporate the comments above and update the code.

**Comment by CannonLock** [comment]
Is the parameter naming finalized here yet?

**Comment by haoming29** [comment]
> SentinelLocation

Let's do `SentinelLocation` for now. I'll revise this PR today but I think `SentinelLocation` sounds good to me

**Comment by CannonLock** [comment]
@haoming29 In testing the UI I hit the following error that killed the web app. 

```
Fatal error occurred at the start of the program. Cleanup started: fail to open SentinelLocation test for StoragePrefix sdfsdf. Directory check failed: stat sdfsdf/test: no such file or directory
```

Notably I used garbage in the config values so not sure if this is really a "bug"

```
origin:
    exports:
        - capabilities:
            - PublicReads
            - Writes
            - Listings
            - DirectReads
            - Reads
          federationprefix: /testtimeusa
          storageprefix: /test0123/cannonlock-test/98y230498q2yt0qwu8gf89q3gt2
        - capabilities:
            - DirectReads
          federationprefix: sdfsdf
          sentinellocation: test
          storageprefix: sdfsdf
```

**Comment by haoming29** [comment]
@CannonLock Do you have a file `test` under `/sdfsdf`?

**Comment by CannonLock** [comment]
@haoming29 No, more so just bummed this took out the web app. Would be nice if we could keep the webapp running through this type of thing so they could continue to edit the config file/read the logs right in the webapp... Just a idea for the future.

**Comment by haoming29** [comment]
> @haoming29 No, more so just bummed this took out the web app. Would be nice if we could keep the webapp running through this type of thing so they could continue to edit the config file/read the logs right in the webapp... Just a idea for the future.

Good point. This would be a corner case where we expect user to have a test file in place before running Pelican. If they decided to change the test file on the go, they should be prepared to have the new test file before making the change. It would be nice if we can have a tooltip/description to emphasize this

File: server_utils/origin.go [hunk 1]
```diff
@@ -246,6 +249,23 @@ func GetOriginExports() (*[]OriginExports, error) {
 	return originExports, nil
 }
 
+func CheckSentinelLocation(exports *[]OriginExport) (ok bool, err error) {
+	for _, export := range *exports {
+		if export.SentinelLocation != "" {
+			sentinelPath := path.Clean(export.SentinelLocation)
+			if path.Base(sentinelPath) != sentinelPath {
+				return false, fmt.Errorf("invalid SentinelLocation path for StoragePrefix %s, file must not contain a directory. Got %s", export.StoragePrefix, export.SentinelLocation)
```
Comment (bbockelm) [inline server_utils/origin.go:257]:
```suggestion
				return false, errors.Errorf("invalid SentinelLocation path for StoragePrefix %s, file must not contain a directory. Got %s", export.StoragePrefix, export.SentinelLocation)
```

File: server_utils/origin.go [hunk 2]
```diff
@@ -246,6 +249,23 @@ func GetOriginExports() (*[]OriginExports, error) {
...
+			}
+			fullPath := filepath.Join(export.StoragePrefix, sentinelPath)
+			_, err := os.Stat(fullPath)
+			if err != nil {
+				return false, errors.Wrap(err, fmt.Sprintf("fail to open SentinelLocation %s for StoragePrefix %s. Directory check failed", export.SentinelLocation, export.StoragePrefix))
```
Comment (bbockelm) [inline server_utils/origin.go:262]:
```suggestion
				return false, errors.Wrapf(err, "fail to open SentinelLocation %s for StoragePrefix %s. Directory check failed", export.SentinelLocation, export.StoragePrefix)
```

**Review by bbockelm — APPROVED** [review]
Two small nitpicks on coding style.  After that, I think this is pre-approved and feel free to merge!
